### PR TITLE
feat: support nric whitelisting 

### PR DESCRIPTION
**Pull request by kevin9foong**
## Problem

Closes FRM-1737
Closes FRM-1774
Closes FRM-1781

## Solution

**Breaking Changes** 
No - this PR is backwards compatible  

## Before & After Screenshots
| Subject | Before | After | 
| ------- | ------- | ----- |
| New whitelist singpass setting | image | image | 
| New whitelist setting with upload | NA | image |
| New not whitelisted error | NA | image | 
| Download whitelist setting csv file secret key input modal | NA | image |

### TC1: Admin can upload a valid whitelisted NRIC list 
- [ ] Create a new storage mode form 
- [ ] Go to settings -> singpass -> enable singpass settings and set to any singpass mode 
- [ ] Click on the 'download a sample csv file' 
- [ ] Modify the 3 sample nric in the csv to 'S9812379B', 'F1612351W', 'S9912370B' and '53342803C'
- [ ] Upload the csv file
- [ ] Set the form to public mode 
- [ ] Respond to the form using only the 'S9812379B' mockpass login 2 times, make sure it is allowed for multiple submissions 
- [ ] Try to respond to the form using any NRIC/FIN not in the whitelist, it should be blocked 
- [ ] Verify in the admin form results page that the responses for the whitelisted NRIC are received. 
- [ ] Go back to the singpass settings page, note that the singpass settings should be greyed out due to form is public. 
- [ ] Try click on delete button of form, it should be disabled. 

### TC2: Admin can download whitelist settings 
- [ ] Click on the download button of whitelist setting, a modal should pop up to request secret key. 
- [ ] Close the modal, it should work 
- [ ] Press download button again, input secret key and click download 
- [ ] Verify csv file is downloaded with 'Respondent' header and correct whitelist submitterId entries ie, 'S9812379B', 'S9912370B', 'F1612351W' and '53342803C'.   

### TC3: Admin can whitelist and use single submission enabled
- [ ] Continuing from the form in TC1, set form to private and turn on the limit to single submission toggle in settings -> singpass
- [ ] Set form to public 
- [ ] Try to respond with  previous 'S9812379B' mockpass login, it should be disallowed since it has submitted previously. 
- [ ] Try to submit 'F1612351W' mockpass login, it should be allowed. 
- [ ] Try to submit for a second time with 'F1612351W' mockpass login, it should be disallowed since it has submitted previously. 
- [ ] Try to submit with a non whitelisted NRIC, it should be disallowed due to not in whitelist. 

### TC4: Admin can remove whitelist 
- [ ] Make form private 
- [ ] Go back to form singpass settings, click delete button. Verify whitelist file is deleted even after refresh. 
- [ ] Try to respond with 'F1612351W' mockpass login, it should still be disallowed due to single submission.
- [ ] Try to respond with S9912370B, it should be allowed.  
- [ ] Try to respond with a mockpass login not previously whitelisted, it should be allowed. 

### TC5: Admin sees appropriate error messages when uploading invalid CSV files
- [ ] Clone and modify the csv file from step 1 by removing the respondent header. Upload it to whitelist setting, we should see the respondent header not present error. 
- [ ] Clone and modify the csv file from step 1 by removing all submitterIds, leaving just the respondents header. Upload it to whitelist setting, we should see the body cannot be empty error.
- [ ] Clone and modify the csv file from step 1 by adding duplicate valid submitterIds. Upload it to whitelist setting, we should see the duplicate present error. 
- [ ] Clone and modify the csv file from step 1 by adding an empty row to middle of the csv file rows. Upload it to whitelist setting, we should see the no empty rows error. 
- [ ] Clone and modify the csv file from step 1 by adding some invalid submitter id such as 'S1234567A'. Upload it to whitelist setting, we should see the invalid id present error with correct example.
- [ ] Upload a csv file that does not have a single column with the header Respondent. we should see a invalid format csv error message 

### TC6: Admin can input file with empty last row without reject. 
- [ ] Add an empty last row to the csv from step 1 using notepad/excel etc. Upload it to whitelist setting. It should not be rejected. 

### TC7: Admin can whitelist for payment form 
- [ ] Create a new storage mode form with payments connected and payment question. 
- [ ] Go to settings -> singpass -> enable singpass settings and set to any singpass mode 
- [ ] Upload the valid csv file from step 1
- [ ] Set the form to public mode 
- [ ] Respond to the form using only the 'S9812379B' mockpass login and make payment, it should be successful 
- [ ] Respond to form using any mockpass login not in whitelist, it should not allow them to enter the form. 
- [ ] Verify in the admin form results page that the responses for the whitelisted NRIC are received. 

### TC8: Admin cannot upload whitelist above 20mb 
- [ ] upload a whitelist that is >20MB, it should fail 

## Deploy Notes
**New Env**:
- `DOWNLOAD_FORM_WHITELIST_RATE_LIMIT`
- `UPLOAD_FORM_WHITELIST_RATE_LIMIT`

**New dependencies**:
- `tweetnacl`: installed on frontend. used for invoking `/shared` directory decryption method.  
- `papaparse`: installed on `frontend`. used for client csv file processing to string. chosen due to good maintenance effort/docs. fastest parser available with fast mode. 
- `multer`: installed on `backend`. used for `multipart/form-data` request body handling for large csv string streaming. express docs recommended package. 

**New dev dependencies**:
- `@types/papaparse`
- `@types/multer`

### CODE REVIEW

File: frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+  downloadFileName: string
+}
+
+const SECRET_KEY_NAME = 'secretKey'
+
+interface SecretKeyFormInputs {
+  [SECRET_KEY_NAME]: string
+}
+
+const useSecretKeyWhitelistFileModal = ({
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx]: /nit can consider moving this to it's own file `useSecretKeyWhitelistFileModal.tsx` in the same folder

File: frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+              </FormControl>
+              <Button
+                rightIcon={<BiRightArrowAlt fontSize="1.5rem" />}
+                type="submit"
+                isFullWidth
+                isDisabled={!isSecretKeyUploaded}
+              >
+                Download
+              </Button>
+            </form>
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx]: Can this reuse the component from `frontend/src/components/SecretKeyVerificationInput/SecretKeyVerificationInput.tsx`?

**Context**
We have additional changes that updates this Input field to support drag n drop in another PR.

File: frontend/src/features/admin-form/settings/mutations.ts [hunk 3]
```diff
@@ -342,6 +357,25 @@ export const useMutateFormSettings = () => {
     },
   )
 
+  // TODO: fix the TS typing
+  const mutateFormWhitelistSetting = useMutation<any, Error, any, any>(
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/mutations.ts]: Can the types be removed? TS should automagically infer them.

File: frontend/src/features/admin-form/settings/queries.ts [hunk 4]
```diff
@@ -31,6 +38,26 @@ export const useAdminFormSettings = (): UseQueryResult<FormSettings> => {
   )
 }
 
+/**
+ * Need to use React query's `refetch` to trigger a fetch.
+ * @precondition Must be wrapped in a Router as `useParam` is used.
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/queries.ts]: Typo here? Doesn't look like we are using `useParam` in the function body.

File: frontend/src/features/public-form/PublicFormProvider.tsx [hunk 5]
```diff
@@ -369,14 +389,19 @@ export const PublicFormProvider = ({
     data?.form.responseMode === FormResponseMode.Encrypt &&
     data.form.payments_field.enabled
 
+  const hasMyInfoError = !!(
+    data?.errorCodes &&
+    data.errorCodes.find((errorCode) => errorCode === ErrorCode.myInfo)
```
Comment (KenLSM) [inline frontend/src/features/public-form/PublicFormProvider.tsx]: /nit `data?.errorCodes.find((errorCode) => errorCode === ErrorCode.myInfo)`

File: frontend/src/features/admin-form/settings/mutations.ts (hunk 3 above)
Comment (kevin9foong) [reply frontend/src/features/admin-form/settings/mutations.ts]: removed

File: frontend/src/features/admin-form/settings/queries.ts (hunk 4 above)
Comment (kevin9foong) [reply frontend/src/features/admin-form/settings/queries.ts]: thanks for the catch, i changed the useQuery logic to use `fetchQuery` but forgot to update the comment!

File: frontend/src/features/public-form/PublicFormProvider.tsx (hunk 5 above)
Comment (kevin9foong) [reply frontend/src/features/public-form/PublicFormProvider.tsx]: @KenLSM , i have simplified it to: 

`const hasMyInfoError = !!data?.errorCodes?.find(
    (errorCode) => errorCode === ErrorCode.myInfo,
  )`

File: frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx (hunk 2 above)
Comment (kevin9foong) [reply frontend/src/features/admin-form/settings/components/AuthSettingsSection/SecretKeyDownloadWhitelistFileModal.tsx]: Resolved by setting common `SecretKeyFormModal` component which uses the reusable `useSecretKeyForm` hook

File: frontend/src/utils/parseCsvFileToCsvString.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import Papa from 'papaparse'
+
+export const parseCsvFileToCsvStringWithoutChunking = (
```
Comment (KenLSM) [inline frontend/src/utils/parseCsvFileToCsvString.ts]: /nit: I would assume that this function would naturally parse csv files "without chunking", as my base assumption. I was initially confused that it is explicitly specified that we will not chunk here.

File: frontend/src/features/admin-form/settings/components/SecretKeyActivationModal.tsx [hunk 7]
```diff
@@ -1,245 +1,38 @@
...
-    </Modal>
+    <SecretKeyFormModal
+      isLoading={isLoading}
+      onClose={onClose}
+      isOpen={isOpen}
+      publicKey={publicKey}
+      modalActionText="Active your form"
+      submitButtonText="Activate form"
+      onSecretKeyFormSubmit={onSecretKeyFormSubmit}
+      hasAck={true}
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/SecretKeyActivationModal.tsx]: /nit can just `hasAck` 

```<FooElement hasAck /> === <FooElement hasAck={true} />```

File: frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+/**
+ * Reusable hook for secret key form logic.
+ * Handles validation of secret key and supports various secret key selection methods such as file upload.
+ */
+export const useSecretKeyForm = ({
+  publicKey,
+  onClose,
+  onSecretKeyFormSubmit,
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx]: /nit generally we don't need too specific with the name. With the appropriate context, it is understood that we are doing a `onSecretKeyFormSubmit`. Thus, a short `onSubmit` would suffice, similar to `onSecretKeyFormClose` vs `onClose`

File: frontend/src/features/admin-form/settings/mutations.ts [hunk 9]
```diff
@@ -342,6 +357,24 @@ export const useMutateFormSettings = () => {
...
+  const mutateFormWhitelistSetting = useMutation(
+    (whitelistCsvString: Promise<string> | null) => {
+      return updateFormWhitelistSetting(formId, whitelistCsvString)
+    },
+    {
+      onSuccess: (_newData, variable) => {
+        generateSuccessToast(
+          variable
+            ? 'Your CSV has been uploaded successfully.'
+            : 'Your CSV has been removed successfully.',
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/mutations.ts:369]: Good that we're optimising for searchability and readability there 👍

File: frontend/src/utils/parseCsvFileToCsvString.ts (hunk 6 above)
Comment (kevin9foong) [reply frontend/src/utils/parseCsvFileToCsvString.ts]: resolved!

File: frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx (hunk 8 above)
Comment (kevin9foong) [reply frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx]: changed to `onSubmit`!

**Comment by kevin9foong** [comment]
Resolved all the PR comments!

**Comment by KenLSM** [comment]
Fixed merge conflicts with `develop` in https://github.com/opengovsg/FormSG/pull/7534/commits/8d8dc8b75a69df4bcfed0a8191a8f69f93d5c9ad. Based off https://github.com/opengovsg/FormSG/pull/7372/files

File: frontend/src/features/admin-form/settings/components/AuthSettingsSection/FormWhitelistAttachmentField.tsx [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    required: true,
+    disabled: isDisabled,
+    fieldType: BasicField.Attachment,
+    attachmentSize: AttachmentSize.TwentyMb,
+  }
+
+  const { publicKey, whitelistedSubmitterIds } = settings
+
+  const isWhitelistEnabled =
+    whitelistedSubmitterIds && whitelistedSubmitterIds.isWhitelistEnabled
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/components/AuthSettingsSection/FormWhitelistAttachmentField.tsx]: /nit can be terser

```suggestion
  const isWhitelistEnabled = whitelistedSubmitterIds?.isWhitelistEnabled
```

File: frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+  onSubmit: ({ secretKey, ack }: SecretKeyFormInputs) => void
+  onClose: () => void
+  hasAck: boolean
+}
+
+/**
+ * Reusable hook for secret key form logic.
+ * Handles validation of secret key and supports various secret key selection methods such as file upload.
+ */
+export const useSecretKeyForm = ({
```
Comment (KenLSM) [inline frontend/src/features/admin-form/settings/hooks/useSecretKeyForm.tsx:24]: Good refactor to extract this hook

File: package.json [hunk 12]
```diff
@@ -209,7 +211,6 @@
     "copyfiles": "^2.4.1",
     "core-js": "^3.28.0",
     "coveralls": "^3.1.1",
-    "csv-parse": "^5.3.6",
```
Comment (KenLSM) [inline package.json:212]: Nice! Removal of unused libraries. A step towards better maintainability.

**Review by KenLSM — APPROVED** [review]
LGTM. Just minor nits. We also have some config that we'll need to update on SSM. We can pair after we've merge this in.

File: frontend/src/features/admin-form/settings/components/AuthSettingsSection/FormWhitelistAttachmentField.tsx (hunk 10 above)
Comment (kevin9foong) [reply frontend/src/features/admin-form/settings/components/AuthSettingsSection/FormWhitelistAttachmentField.tsx]: fixed!

**Comment by kevin9foong** [comment]
Did a final sanity check to ensure all is working as expected. Squash and merging now

**Pull request merged**